Add to HealthMetadata information about ShardLimits

HiDAl opened the pull request:
create all the machinery to have the shard limits information from the master node. Later, these values will be used by the new shard limits health indicator.   

* adds a new `ShardLimitsMetadata` object to be filled up by the master node and used by the designed `health-node`
* wire up the listeners for changes of shard limits-related settings


issue #91119

Commit:
new record object to store shardLimits information

Commit:
add to `HealthMetadata` the new `ShardLimits` info

Commit:
add tests for shard limits changes

Commit:
add listeners for shard limits setting changes

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

HiDAl committed:
Update docs/changelog/94116.yaml

HiDAl commented:
@elasticsearchmachine run elasticsearch-ci/bwc

Commit:
mark the minimum version since when the metadata is available

HiDAl commented:
@elasticsearchmachine run elasticsearch-ci/bwc

Commit:
fix versioning of shardLimits read/write

Commit:
spotless

andreidan commented on server/src/internalClusterTest/java/org/elasticsearch/health/HealthMetadataServiceIT.java:
```diff
@@ -40,33 +48,43 @@ public void testEachMasterPublishesTheirThresholds() throws Exception {
                 ByteSizeValue customMaxHeadroom = percentageMode ? randomBytes : ByteSizeValue.MINUS_ONE;
-                String nodeName = startNode(internalCluster, customWatermark, customMaxHeadroom.toString());
+                var customShardLimits = new HealthMetadata.ShardLimits(randomIntBetween(0, 1000), randomIntBetween(1001, 2000));
+                String nodeName = startNode(internalCluster, customWatermark, customMaxHeadroom.toString(), customShardLimits);
                 watermarkByNode.put(nodeName, customWatermark);
                 maxHeadroomByNode.put(nodeName, customMaxHeadroom);
+                shardLimitsPerNode.put(nodeName, customShardLimits);
+
             }
             ensureStableCluster(numberOfNodes);
 
             String electedMaster = internalCluster.getMasterName();
             {
-                HealthMetadata.Disk diskMetadata = HealthMetadata.getFromClusterState(internalCluster.clusterService().state())
-                    .getDiskMetadata();
+                var healthMetadata = HealthMetadata.getFromClusterState(internalCluster.clusterService().state());
+                var diskMetadata = healthMetadata.getDiskMetadata();
```
We should leave these now, but for the future let's please have these cosmetics as a separate PR as it makes reviewing a bit murkier when they're baked in with business logic changes.

HiDAl replied:
sure!

andreidan commented on server/src/internalClusterTest/java/org/elasticsearch/health/HealthMetadataServiceIT.java:
```diff
@@ -22,6 +22,14 @@
 import java.util.Map;
 import java.util.Set;
 
+import static org.elasticsearch.cluster.routing.allocation.DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_MAX_HEADROOM_SETTING;
+import static org.elasticsearch.cluster.routing.allocation.DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_WATERMARK_SETTING;
+import static org.elasticsearch.cluster.routing.allocation.DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_MAX_HEADROOM_SETTING;
+import static org.elasticsearch.cluster.routing.allocation.DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_WATERMARK_SETTING;
+import static org.elasticsearch.cluster.routing.allocation.DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_LOW_DISK_MAX_HEADROOM_SETTING;
+import static org.elasticsearch.cluster.routing.allocation.DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_LOW_DISK_WATERMARK_SETTING;
```
We should leave these now, but for the future let's please have these cosmetics as a separate PR as it makes reviewing a bit murkier when they're baked in with business logic changes.

andreidan commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadata.java:
```diff
@@ -84,22 +97,95 @@ public Disk getDiskMetadata() {
+    }
+
+    /**
+     * Contains the thresholds needed to determine the health of a cluster from the amount of room available to create new shards. These
```
```suggestion
     * Contains the thresholds needed to determine the health of a cluster when it comes to the amount of room available to create new shards. These
```

andreidan commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadata.java:
```diff
@@ -84,22 +97,95 @@ public Disk getDiskMetadata() {
+
+            private Builder() {}
+
+            private Builder(ShardLimits shardLimits) {
+                this.maxShardPerNode = shardLimits.maxShardsPerNode;
+                this.maxShardPerNodeFrozen = shardLimits.maxShardsPerNodeFrozen;
+            }
+
+            public Builder maxShardPerNode(int maxShardPerNode) {
+                this.maxShardPerNode = maxShardPerNode;
+                return this;
+            }
+
+            public Builder maxShardPerNodeFrozen(int maxShardPerNodeFrozen) {
+                this.maxShardPerNodeFrozen = maxShardPerNodeFrozen;
+                return this;
+            }
```
Shall we validate the values are `>= 1` ?

andreidan commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadata.java:
```diff
@@ -84,22 +97,95 @@ public Disk getDiskMetadata() {
+     */
+    public record ShardLimits(int maxShardsPerNode, int maxShardsPerNodeFrozen) implements ToXContentFragment, Writeable {
+
+        public static final ShardLimits UNKNOWN = new ShardLimits(-1, -1);
```
This special value means we now have to deal with 3 states:

- unknown
- actual values
- missing limits ( `null` )

Could the `unknown` special value be eliminated and treat `null` as unknown? 
This way we'll have to deal with 2 values logic.

HiDAl replied:
makes sense, will convert them to `null and actual values`

andreidan commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java:
```diff
@@ -82,40 +85,39 @@ private void registerListeners() {
-        clusterSettings.addSettingsUpdateConsumer(
+            CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_FROZEN_WATERMARK_SETTING
+        )
+            .forEach(
+                setting -> clusterSettings.addSettingsUpdateConsumer(
+                    setting,
+                    value -> updateOnSettingsUpdated(new UpdateDiskMetadata(setting.getKey(), value.getStringRep()))
+                )
+            );
```
Is this the same as the below `forEach` ? 

If they are can we just have one `Stream.of` for all disk related values?

HiDAl replied:
Sadly no, they are handling `ByteSizeValue` and `RelativeByteSizeValue`, which aren't sharing the same ancestor, so it's not possible to abstract into 1 stream :(

andreidan replied:
Makes sense, thanks :) (my eyes glanced over that difference)

andreidan commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java:
```diff
@@ -182,7 +183,15 @@ public void onFailure(@Nullable Exception e) {
         }
 
-        abstract ClusterState execute(ClusterState currentState);
+        ClusterState execute(ClusterState currentState) {
+            var initialHealthMetadata = HealthMetadata.getFromClusterState(currentState);
+            var finalHealthMetadata = doExecute(initialHealthMetadata);
+            return finalHealthMetadata.equals(initialHealthMetadata)
+                ? currentState
+                : currentState.copyAndUpdate(b -> b.putCustom(HealthMetadata.TYPE, finalHealthMetadata));
+        }
+
+        abstract HealthMetadata doExecute(HealthMetadata initialHealthMetadata);
```
This is neat ❤️

andreidan reviewed (commented):
Thanks Pablo, this generally looks great. 

I left a few suggestions

gmarouli commented on server/src/internalClusterTest/java/org/elasticsearch/health/HealthMetadataServiceIT.java:
```diff
@@ -40,33 +48,43 @@ public void testEachMasterPublishesTheirThresholds() throws Exception {
                 watermarkByNode.put(nodeName, customWatermark);
                 maxHeadroomByNode.put(nodeName, customMaxHeadroom);
+                shardLimitsPerNode.put(nodeName, customShardLimits);
+
```
```suggestion
```

gmarouli commented on server/src/test/java/org/elasticsearch/health/metadata/HealthMetadataSerializationTests.java:
```diff
@@ -58,7 +58,11 @@ protected ClusterState.Custom mutateInstance(ClusterState.Custom instance) {
+    }
+
+    private static HealthMetadata.ShardLimits randomShardLimitsMetadata() {
+        return new HealthMetadata.ShardLimits(randomInt(), randomInt());
```
I am not 100% sure but I think `randomInt()` generates negative values too. So, could you double check this to avoid adding a flaky test when the validation for only positive numbers is added.

gmarouli commented on server/src/test/java/org/elasticsearch/health/metadata/HealthMetadataSerializationTests.java:
```diff
@@ -105,8 +109,19 @@ static HealthMetadata.Disk mutateDiskMetadata(HealthMetadata.Disk base) {
+        var maxShardPerNode = base.maxShardsPerNode();
+        var maxShardPerNodeFrozen = base.maxShardsPerNodeFrozen();
+        switch (randomInt(1)) {
+            case 0 -> maxShardPerNode = randomInt();
```
Same here.

gmarouli reviewed (commented):
Thank for working on this, it's nice to see the `HealthMetadata` being extended :). I left some minor comments and I second Andrei's comments :).

Commit:
tackle PR comments

andreidan commented:
`elasticsearch-ci/bwc` failure seems transient. Restarting

andreidan commented:
@elasticmachine run elasticsearch-ci/bwc

andreidan reviewed (approved):
LGTM, thanks for iterating on this Pablo

HiDAl commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into new-shard-limit-metadata

andreidan commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadata.java:
```diff
@@ -84,22 +97,116 @@ public Disk getDiskMetadata() {
+            out.writeInt(maxShardsPerNodeFrozen);
+        }
+
+        public boolean isUnknown() {
+            return maxShardsPerNode != -1 && maxShardsPerNodeFrozen != -1;
+        }
```
I believe this shouldn't be needed anymore? A missing (`null`) ShardLimits represents `unknown`

Commit:
remove unused method
fix typo on method/variable names

Commit:
move the validation to the Builders

ShardLimits validation shouldn't run as part of the construction of the
object, otherwise it will block the serialization path.

andreidan commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadata.java:
```diff
@@ -72,6 +82,9 @@ public Iterator<? extends ToXContent> toXContentChunked(ToXContent.Params ignore
             builder.startObject(DISK_METADATA.getPreferredName());
             diskMetadata.toXContent(builder, params);
             builder.endObject();
+            builder.startObject(SHARD_LIMITS_METADATA.getPreferredName());
+            shardLimitsMetadata.toXContent(builder, params);
+            builder.endObject();
```
this needs a null check

Commit:
ensure that `shardLimitsMetadata` can be null

when serializing the value to the wire, we ensure that it's writable/readable using Optional methods.
add proper tests to check the null case

HiDAl commented:
executed the serialization test 1000 times, no errors:

```bash
➜  elasticsearch git:(new-shard-limit-metadata) ✗ ./gradlew ':server:test' --tests "org.elasticsearch.health.metadata.HealthMetadataSerializationTests" -Dtests.iters=1000
=======================================
...
=======================================

> Task :server:test
[...]
BUILD SUCCESSFUL in 18s
61 actionable tasks: 2 executed, 59 up-to-date

Publishing build scan...
https://gradle-enterprise.elastic.co/s/endahupjxcp4y
```

Commit:
remove unneeded tests

HiDAl commented:
@elasticmachine run elasticsearch-ci/part-2
